database/storage_engine.h

```diff
@@ -18,6 +28,12 @@ struct storage_engine {
     RRD_MEMORY_MODE id;
     const char* name;
     STORAGE_ENGINE_API api;
+    STORAGE_ENGINE_INSTANCE* context;
```

Comment:
A storage engine can have multiple instances. I think the proper way to handle global vs. per-host instances is to have a `context` pointer in `storage_engine_instance`. This will work as long as:

1. The first host that will be created is localhost,
2. The last host that will be deleted is localhost,
3. A localhost host outlives all other hosts.

AFAICT, this assumptions are true today, which means we can do global initialization of a storage_engine when we create the 1st storage_engine_instance of a localhost `RRDHOST`. 

As an example, the storage_engine_instance for dbengine could look like this:

```
struct storage_engine_instance {
    STORAGE_ENGINE* engine;
    void *context
};

struct dbengine_storage_engine_instance_context {
   RRDHOST *host;
   rrdengine_instance *multidb_ctx;
   rrdengine_instance *rrdeng_ctx;
};

...
...
...

STORAGE_ENGINE_INSTANCE* dbengine_storage_engine_new(STORAGE_ENGINE* engine, RRDHOST *host) {
   struct dbengine_storage_engine_instance_context *dbe_ctx = callocz(1, ...);

   dbe_ctx->host = host;

   if (host == localhost) {
       // init multidb_ctx
       ...
       dbe_ctx->multidb_ctx = ...;
   }

   if (is_legacy_child(host) {
       // init regular rrdeng_ctx for host
       ...
       rrdeng_init(..., &dbe_ctx->rrdeng_ctx, ...);
      ...
   }

   host->storage_engine_instance = callocz(1, ...);
   host->storage_engine_instance->context = dbe_ctx;
}
```

What do you think?

Reply:
As discussed last week, the per-engine global context is necessary, for instance to own the global state of the engine (TCP connection to the DB etc). Following previous discussions and reviews, global engine states are now only initialized when necessary.